### ISSUE TITLE
Overwritten tensor by `DataLoader` with `device='mps'`

### ISSUE THREAD

**Issue by PierreGtch** [pytorch/pytorch#93356]
### 🐛 Describe the bug

The tensors given to a dataloader are overwritten to the first value of the tensor if it is unidimensional:
```python
import torch
from torch.utils.data import TensorDataset, DataLoader

list(DataLoader(TensorDataset(torch.tensor([3,2]).to('mps')), shuffle=True, batch_size=2))
```
gives:
```python
[[tensor([3, 3], device='mps:0')]]
```

Other undesired  examples:
```python
In [58]: list(DataLoader(TensorDataset(torch.tensor([3.1,2]).to('mps')), shuffle=True, batch_size=2)) # with floats
Out[58]: [[tensor([3.1000, 3.1000], device='mps:0')]]

In [59]: list(DataLoader(TensorDataset(torch.tensor([3,2]).to('mps')), shuffle=True, batch_size=1)) # batch size of 1
Out[59]: [[tensor([3], device='mps:0')], [tensor([3], device='mps:0')]]

In [60]: list(DataLoader(TensorDataset(torch.tensor([3,2]).to('mps')), shuffle=False, batch_size=1)) # non shuffled and batch size of 1
Out[60]: [[tensor([3], device='mps:0')], [tensor([3], device='mps:0')]]

In [61]: list(DataLoader(TensorDataset(torch.tensor([3,2]).to('mps')), shuffle=False, batch_size=2)) # non shuffled
Out[61]: [[tensor([3, 3], device='mps:0')]]
```

However, if the tensor has multiple dimensions, it works fine:
```python
In [56]: list(DataLoader(TensorDataset(torch.tensor([[1,2],[3,4]]).to('mps')), shuffle=True, batch_size=2))
Out[56]:
[[tensor([[3, 4],
          [1, 2]], device='mps:0')]]

In [57]: list(DataLoader(TensorDataset(torch.tensor([[1,2],[3,4]]).to('mps')), shuffle=True, batch_size=1))
Out[57]: [[tensor([[1, 2]], device='mps:0')], [tensor([[3, 4]], device='mps:0')]]
```

### Versions

```
PyTorch version: 2.0.0.dev20230117
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.22.4
Libc version: N/A

Python version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:13) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-lightning==1.8.1
[pip3] torch==2.0.0.dev20230117
[pip3] torchdata==0.6.0.dev20230118
[pip3] torchmetrics==0.11.0
[conda] numpy                     1.24.1          py310h5d7c261_0    conda-forge
[conda] pytorch                   2.0.0.dev20230117        py3.10_0    pytorch-nightly
[conda] pytorch-lightning         1.8.1              pyhd8ed1ab_0    conda-forge
[conda] torchdata                 0.6.0.dev20230118           py310    pytorch-nightly
[conda] torchmetrics              0.11.0             pyhd8ed1ab_0    conda-forge
```

cc @SsnL @VitalyFedyunin @ejguan @NivekT @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by ejguan** [comment]
I think it's a duplicated issue https://github.com/pytorch/pytorch/issues/92265 and https://github.com/pytorch/pytorch/issues/91393

This is about `torch.stack` in DataLoader's default_collate function to stack one-dimensional mps Tensor. Removing the labels of DataLoader and Data.

**Comment by PierreGtch** [comment]
Ah yes indeed, sorry
